torch/csrc/dynamo/eval_frame.c

```diff
@@ -663,11 +830,18 @@ static PyObject* _custom_eval_frame(
     return eval_frame_default(tstate, frame, throw_flag);
   }
 
-  CacheEntry* extra = get_cache_entry(frame->f_code);
+  ExtraState* extra = get_extra_state(frame->f_code);
   if (extra == SKIP_CODE || (callback == Py_False && extra == NULL)) {
     DEBUG_TRACE("skip %s", get_frame_name(frame));
     return eval_frame_default(tstate, frame, throw_flag);
   }
+  CacheEntry* cache_entry = extract_cache_entry(extra);
+  FrameState* frame_state = extract_frame_state(extra);
+  if (frame_state == NULL) {
+    // frame_state becomes a new reference. It gets cleaned up by
+    // set_extra_state or skip later.
+    frame_state = PyDict_New();
```

Comment:
Ugh. Why not just immediately set it into extra? As written here, the variable `frame_state` is either owning or non-owning, depending on whether or not there was a frame state on extra. That way lies madness. Just populate it immediately on access.

Reply:
Introduced a new function - init_and_set_extra_state which inits the extra_state with (CacheEntry=NULL, FrameState=PyDict). This resolves the owning or non-owning depending on the code path.